theseus/embodied/motionmodel/misc.py

```diff
@@ -0,0 +1,160 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Optional, Sequence, Tuple, Union
+
+import torch
+
+from theseus.core import CostFunction, CostWeight, Variable, as_variable
+from theseus.geometry import SE2, Point3, Vector
+
+
+class HingeLoss(CostFunction):
+    def __init__(
+        self,
+        vector: Vector,
+        limit: Union[float, torch.Tensor, Variable],
+        threshold: Union[float, torch.Tensor, Variable],
+        cost_weight: CostWeight,
+        name: Optional[str] = None,
+        side: str = "both",
+        dims: Optional[Sequence[int]] = None,
+    ):
+        super().__init__(cost_weight, name=name)
+        self.vector = vector
+        self.limit = as_variable(limit, name=f"{self.name}__vlimit")
+        self.threshold = as_variable(threshold, name=f"{self.name}__vthres")
+        for v in [self.limit, self.threshold]:
+            if not v.ndim == 2 and v.shape[1] == 1:
+                raise ValueError("Limit and threshold must be scalar variables.")
+        if self.threshold.tensor.max() < 0.0:
+            raise ValueError("The threshold must be a positive scalar.")
+        self.register_optim_var("vector")
+        self.register_aux_vars(["limit", "threshold"])
+        if side not in ["below", "above", "both"]:
+            raise ValueError("side must be one of 'both', 'above', 'below'.")
+        self.side = side
+        self.dims = dims
+
+    def dim(self):
+        if self.dims is not None:
+            return len(self.dims)
+        return self.vector.dof()
+
+    def _compute_error(self) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        vector = self.vector.tensor
+        limit = self.limit.tensor
+        threshold = self.threshold.tensor
```

Comment:
Each dim in vector has it's own limit and threshold that could be same or different than other dims, right?

I am guessing if you wanted to not apply any hinge to a specific dimension you could just pass a inf/max or something for limit and threshold for that dim.

Reply:
In the current implementation we force `limit` and `threshold` to be scalars.

Reply:
Ah so it's the same value for all dims. We will likely need to extend this when we start working on full body planning and joint limits get involved. But good for now to make progress.